bentoml/marshal/marshal.py

```diff
@@ -12,136 +12,159 @@
+                name=service_name + '_mb_requestmb_duration_seconds',
+                documentation=service_name + "API HTTP request duration in seconds",
+                namespace=namespace,
+                labelnames=['endpoint', 'http_response_code'],
+            )
+            self.metrics_request_in_progress = Gauge(
+                name=service_name + "_mb_request_in_progress",
+                documentation='Totoal number of HTTP requests in progress now',
+                namespace=namespace,
+                labelnames=['endpoint', 'http_method'],
+            )
+            self.metrics_request_exception = Counter(
+                name=service_name + "_mb_request_exception",
+                documentation='Totoal number of service exceptions',
+                namespace=namespace,
+                labelnames=['endpoint', 'exception_class'],
+            )
+            self.metrics_request_total = Counter(
+                name=service_name + "_mb_request_total",
+                documentation='Totoal number of service exceptions',
+                namespace=namespace,
+                labelnames=['endpoint', 'http_response_code'],
+            )
 
-        async def _func(inputs):
-            parade = self.get_parade()
-            _id = parade.feed(inputs)
-            async with parade.returned:
-                await parade.returned.wait()
-            return parade.batch_output[_id]
+        async def request_dispatcher(self, request):
+            func = super(_MarshalService, self).request_dispatcher
+            api_name = request.match_info["name"]
+            _metrics_request_in_progress = self.metrics_request_in_progress.labels(
+                endpoint=api_name, http_method=request.method,
+            )
+            _metrics_request_in_progress.inc()
+            time_st = time.time()
+            try:
+                resp = await func(request)
+            except Exception as e:  # pylint: disable=broad-except
+                self.metrics_request_exception.labels(
+                    endpoint=api_name, exception_class=e.__class__.__name__
+                ).inc()
+                resp = aiohttp.web.Response(status=500)
+            self.metrics_request_total.labels(
+                endpoint=api_name, http_response_code=resp.status
+            ).inc()
+            self.metrics_request_duration.labels(
+                endpoint=api_name, http_response_code=resp.status
+            ).observe(time.time() - time_st)
+            _metrics_request_in_progress.dec()
+            return resp
+
+        async def _batch_handler_template(self, requests, api_name):
+            func = super(_MarshalService, self)._batch_handler_template
+            self.metrics_request_batch_size.labels(endpoint=api_name).observe(
+                len(requests)
+            )
+            return await func(requests, api_name)
 
-        return _func
+    return _MarshalService
 
 
+@metrics_patch
```

Comment:
I remember we discussed this, do we still want to add a config to disable/enable this?  something like`@metrics_patch(conf['marshal']['enable_metrics'])` here?